### ISSUE TITLE
Unresolved dependency: proguard

### ISSUE THREAD

**Issue by gpolya** [jvoegele/gradle-android-plugin#2]
When I try to build a freshly created project as described in the [wiki](http://github.com/jvoegele/gradle-android-plugin/wiki), I get the following error:

> Execution failed for task ':proguard'.  
> Cause: Could not resolve all dependencies for configuration 'proguard':  
> - unresolved dependency: net.sf.proguard#proguard;4.4: not found

It's the first time I'm using Gradle. The gradle cache contains only the android-plugin (which was downloaded when building). I tried both Gradle 0.9-rc-1 and 0.8.

Skipping ProGuard with "proguard.enabled = false" doesn't seem to work either. If I do that, the build fails at

> :androidPackageDebug  
> [ant:apply] no input files specified

It seems like [androidPackageDebug expects files in the ProGuard output directory](http://github.com/jvoegele/gradle-android-plugin/blob/master/src/main/groovy/com/jvoegele/gradle/plugins/android/AndroidPlugin.groovy#L229).

**Comment by jvoegele** [comment]
The ProGuard problem can be resolved by adding the following to your build.gradle file:

repositories {
    mavenCentral()
}

I will add this information to the wiki page.

However, even with this fix in place, I'm currently encountering a NullPointerException for some reason.  I'll look into this problem, but let me know if you also encounter it.

**Comment by gpolya** [comment]
I added mavenCentral() and the ProGuard part is working now. But I also get the NullPointerException now. The [solution to this stackoverflow question](http://stackoverflow.com/questions/3799368/) might help fixing this problem.

**Comment by thevery** [comment]
jvoegele, gpolya, 
have you managed to resolve NPE problem?

**Comment by fdasoghe** [comment]
Hello.

I'm getting into the NPE problem too.
I followed the instruction in the wiki, creating an Eclipse project using the ADT.
But when I run gradle with task androidPackageDebug I get the NPE:

org.gradle.api.tasks.LocationAwareTaskExecutionException: Build file 'D:\dev\workspace-3.6.1-gradle\prova1\build.gradle'
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter] Execution failed for task ':androidPackageDebug'.
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.tasks.DefaultTaskExecuter.executeActions(DefaultTaskExecuter.java:66)
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.tasks.DefaultTaskExecuter.execute(DefaultTaskExecuter.java:43)
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.project.taskfactory.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:32)
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.project.taskfactory.ExecutionShortCircuitTaskExecuter.execute(ExecutionShortCircuitTaskExecuter.java:50)
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.tasks.SkipTaskExecuter.doExecute(SkipTaskExecuter.java:57)
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.tasks.SkipTaskExecuter.execute(SkipTaskExecuter.java:35)
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.tasks.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:32)
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.AbstractTask.execute(AbstractTask.java:231)
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.execution.DefaultTaskGraphExecuter.executeTask(DefaultTaskGraphExecuter.java:167)
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.execution.DefaultTaskGraphExecuter.doExecute(DefaultTaskGraphExecuter.java:160)
16:50:55.143 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.execution.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:78)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.execution.TaskNameResolvingBuildExecuter.execute(TaskNameResolvingBuildExecuter.java:113)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.execution.DelegatingBuildExecuter.execute(DelegatingBuildExecuter.java:54)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:153)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:107)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:75)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.launcher.CommandLineActionFactory$RunBuildAction.run(CommandLineActionFactory.java:171)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.launcher.CommandLineActionFactory$WithLoggingAction.run(CommandLineActionFactory.java:192)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.launcher.Main.execute(Main.java:51)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.launcher.Main.main(Main.java:38)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.launcher.GradleMain.main(GradleMain.java:52)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter] Caused by: : java.lang.NullPointerException
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:116)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.project.ant.BasicAntBuilder.doInvokeMethod(BasicAntBuilder.java:74)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.project.DefaultAntBuilder.super$3$invokeMethod(DefaultAntBuilder.groovy)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.project.DefaultAntBuilder.invokeMethod(DefaultAntBuilder.groovy:37)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at com.jvoegele.gradle.plugins.android.AndroidPlugin.androidPackage(AndroidPlugin.groovy:242)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at com.jvoegele.gradle.plugins.android.AndroidPlugin.this$2$androidPackage(AndroidPlugin.groovy)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at com.jvoegele.gradle.plugins.android.AndroidPlugin$_defineAndroidPackageDebugTask_closure4.doCall(AndroidPlugin.groovy:113)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:445)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:432)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.gradle.api.internal.tasks.DefaultTaskExecuter.executeActions(DefaultTaskExecuter.java:58)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    ... 20 more
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter] Caused by: java.lang.NullPointerException
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at com.android.ant.ApkBuilderTask.execute(ApkBuilderTask.java:239)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:291)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
16:50:55.159 [ERROR] [org.gradle.BuildExceptionReporter]    ... 29 more

I've read the solution on stackoverflow but I don't understand how to use it. I should configure the &lt;apkbuilder&gt; Ant task but I don't have any Ant task to configure, of course... so what can I do?

Thanks for any help, it would be much appreciated as this seems to be a great tool to use!

**Comment by fdasoghe** [comment]
I forgot to say: I'm using latest version of tools at the time of writing:
- Eclipse 3.6.1<br>
- ADT 0.9.9<br>
- Android SDK r07<br>
  and I'm building for Android Platform 2.1.

Cheers,

think01

**Comment by fdasoghe** [comment]
Hello again.

I digged a bit into your sources, in file AndroidPlugin.groovy.

It seems you have the problem pointed out in stackoverflow question cited above (http://stackoverflow.com/questions/3799368/).

I get on my console these two warnings:

[INFO] [system.out] WARNING: Using deprecated 'basename' attribute in ApkBuilderTask.Use 'apkfilepath' (path) instead.<br>
[INFO] [system.out] WARNING: Using deprecated &lt;file&gt; inner element in ApkBuilderTask.Use &lt;dex path=...&gt; instead.

Reading the source, in the AndroidPlugin.androidPackage() method you wrote:

```
ant.apkbuilder(outfolder: project.buildDir,
    basename: project.name,
    signed: sign,
    'verbose': verbose) {

      ant.file(path: androidConvention.intermediateDexFile.absolutePath)

      //sourcefolder(path: project.sourceSets.main.java)
      nativefolder(path: androidConvention.nativeLibsDir)
      //jarfolder(path: androidConvention.nativeLibsDir)
    }
```

I think the problem is there. Using the dex element should fix it (for Android SDK r07. For previous versions I'm getting other issues, but probably is my fault and anyway, they are getting obsolete).

Would it be possible to have this fix any time soon? I'm totally out of Groovy development :-(

Thanks so much, cheers,

think01

**Comment by fdasoghe** [comment]
Hello.

Using yours (jvoegele) build instruction I've checked out the trunk version of the plugin.

The error changes a little: now I get a NPE in another line, ApkBuilderTask.java:259

Anyway, I debugged a bit and I did some modification to AndroidPlugin.groovy, the most important one being adding the line: 

```
    resourcefile: androidConvention.intermediateDexFile.getName(),
```

in line 245.

This makes the plugin working perfect for me!

If I come out how to create a Git patch, I can send it to you for review, if you like.

Of course this patch has the drawback of being incompatible with Android SDK versions lesser then r07, but it should be quite easy to support more then one version (maybe with derived classes from the AndroidPlugin base class...).

Cheers,

think01

**Comment by jvoegele** [comment]
I've done something similar in my local repo, except that I've broken the apkbuilder task (and some others) into classes of their own instead of being embedded into the AndroidPlugin class directly.  This way there can be a separate apkbuilder task implementation for different versions of the Android SDK, meaning that applying the fix will not break things for users of older SDK versions.  However, I haven't yet figured out how to determine the SDK version that is in use from the Gradle AndroidPlugin.  Does anyone know of a property to lookup, or some other mechanism to query the SDK version from Ant/Gradle?

**Comment by fdasoghe** [comment]
I've not found a "certified and documented" way.... but it seems you can find the version number in file tools/source.properties (from the root of the SDK installation), in property
    Pkg.Revision

Your solution seems very good, IMHO. I can't wait to have the new plugin version :-)

Keep up the good work!

think01

**Comment by jvoegele** [comment]
I have just pushed a commit that resolves the NullPointerException, and which works on both r6 and r7 revisions of the Android SDK tools.  If you pull the latest changes and build your own copy of the plugin it should work fine for you now.

I have a couple of other items that I want to tackle, and after I get those completed I intend to create the 1.0 version of the gradle-android-plugin.  The 1.0 version will include the NPE fix as well as a few other enhancements.

**Comment by fdasoghe** [comment]
Thanks a lot! I'll grab the trunk version. Of course we can't wait for the 1.0 version :o)

**Comment by jvoegele** [comment]
This issue has been fixed in the just-released 0.9.4 version, so closing out this issue.  The fix will, of course, also be included in the upcoming 1.0 version.
